tests/conftest.py

```diff
@@ -750,3 +760,283 @@ def proxy(request: pytest.FixtureRequest) -> str:
+        ],
+        "priority": [
+            # Ensure we check for a missing metadata file for wheels.
+            FakePackage(
+                "priority",
+                "1.0",
+                "priority-1.0-py2.py3-none-any.whl",
+                MetadataKind.NoFile,
+            ),
+        ],
+        "requires-simple-extra": [
+            # Metadata name is not canonicalized.
+            FakePackage(
+                "requires-simple-extra",
+                "0.1",
+                "requires_simple_extra-0.1-py2.py3-none-any.whl",
+                MetadataKind.Sha256,
+                metadata_name="Requires_Simple.Extra",
+            ),
+        ],
+    }
+
+
+@pytest.fixture(scope="session")
+def html_index_for_packages(
+    shared_data: TestData,
+    fake_packages: Dict[str, List[FakePackage]],
+    tmpdir_factory: pytest.TempPathFactory,
+) -> Path:
+    """Generate a PyPI HTML package index within a local directory pointing to
+    synthetic test data."""
+    html_dir = tmpdir_factory.mktemp("fake_index_html_content")
+
+    # (1) Generate the content for a PyPI index.html.
+    pkg_links = "\n".join(
+        f'    <a href="{pkg}/index.html">{pkg}</a>' for pkg in fake_packages.keys()
+    )
+    index_html = f"""\
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta name="pypi:repository-version" content="1.0">
+    <title>Simple index</title>
+  </head>
+  <body>
+{pkg_links}
+  </body>
+</html>"""
+    # (2) Generate the index.html in a new subdirectory of the temp directory.
+    (html_dir / "index.html").write_text(index_html)
+
+    # (3) Generate subdirectories for individual packages, each with their own
+    # index.html.
+    for pkg, links in fake_packages.items():
+        pkg_subdir = html_dir / pkg
+        pkg_subdir.mkdir()
+
+        download_links: List[str] = []
+        for package_link in links:
+            # (3.1) Generate the <a> tag which pip can crawl pointing to this
+            # specific package version.
+            download_links.append(
+                f'    <a href="{package_link.filename}" {package_link.generate_additional_tag()}>{package_link.filename}</a><br/>'  # noqa: E501
+            )
+            # (3.2) Copy over the corresponding file in `shared_data.packages`.
+            shutil.copy(
+                shared_data.packages / package_link.filename,
+                pkg_subdir / package_link.filename,
+            )
+            # (3.3) Write a metadata file, if applicable.
+            if package_link.metadata != MetadataKind.NoFile:
+                with open(pkg_subdir / package_link.metadata_filename(), "wb") as f:
+                    f.write(package_link.generate_metadata())
+
+        # (3.4) After collating all the download links and copying over the files,
+        # write an index.html with the generated download links for each
+        # copied file for this specific package name.
+        download_links_str = "\n".join(download_links)
+        pkg_index_content = f"""\
```

Comment:
Please consider indenting these and using textwrap.dedent -- completely changing the indentation like this breaks editors that rely on indentation to determine collapse spaces for Python code.